File: apps/api/app/billing/quotes.py
from dataclasses import dataclass
from math import ceil

from .settings import BillingSettings, TaskBudget

INDEPENDENT_QUOTE = "independent_evaluation"
EMPLOYER_QUOTE = "employer_resume"


class UnknownQuoteKindError(Exception):
	pass


@dataclass(frozen=True)
class PointQuote:
	kind: str
	line_items: tuple[TaskBudget, ...]
	cost_ceiling_points: int
	minimum_points: int
	points: int


def point_quote(kind: str, settings: BillingSettings) -> PointQuote:
	if kind == INDEPENDENT_QUOTE:
		budgets = settings.independent_budgets
		minimum = settings.minimum_independent_evaluation_points
	elif kind == EMPLOYER_QUOTE:
		budgets = settings.employer_budgets
		minimum = settings.minimum_employer_resume_points
	else:
		raise UnknownQuoteKindError(kind)
	cost_usd = sum(
		budget.max_input_tokens
		* settings.price_ceiling_usd_per_million_input
		/ 1_000_000
		+ budget.max_output_tokens
		* settings.price_ceiling_usd_per_million_output
		/ 1_000_000
		for budget in budgets
	)
	cost_ceiling_points = ceil(cost_usd * settings.points_per_usd)
	return PointQuote(
		kind=kind,
		line_items=budgets,
		cost_ceiling_points=cost_ceiling_points,
		minimum_points=minimum,
		points=max(minimum, cost_ceiling_points),
	)


def settle_points(reported_cost_usd: float | None, kind: str, settings: BillingSettings) -> int:
	"""Completion charge: the greater of the minimum or the ceiling-rounded reported cost."""

	minimum = (
		settings.minimum_independent_evaluation_points
		if kind == INDEPENDENT_QUOTE
		else settings.minimum_employer_resume_points
	)
	if reported_cost_usd is None:
		return minimum
	return max(minimum, ceil(reported_cost_usd * settings.points_per_usd))

File: apps/api-new/src/billing/quotes.ts
export const INDEPENDENT_QUOTE = "independent_evaluation";
export const EMPLOYER_QUOTE = "employer_resume";

export class UnknownQuoteKindError extends Error {}

export interface TaskBudget {
	task: string;
	maxInputTokens: number;
	maxOutputTokens: number;
}

export interface BillingSettings {
	pointsPerUsd: number;
	minimumIndependentEvaluationPoints: number;
	minimumEmployerResumePoints: number;
	priceCeilingUsdPerMillionInput: number;
	priceCeilingUsdPerMillionOutput: number;
	independentBudgets: readonly TaskBudget[];
	employerBudgets: readonly TaskBudget[];
}

export interface PointQuote {
	kind: string;
	lineItems: readonly TaskBudget[];
	costCeilingPoints: number;
	minimumPoints: number;
	points: number;
}

export const pointQuote = (kind: string, settings: BillingSettings): PointQuote => {
	const independent = kind === INDEPENDENT_QUOTE;
	if (!independent && kind !== EMPLOYER_QUOTE) throw new UnknownQuoteKindError(kind);
	const budgets = independent ? settings.independentBudgets : settings.employerBudgets;
	const minimum = independent ? settings.minimumIndependentEvaluationPoints : settings.minimumEmployerResumePoints;
	const costUsd = budgets.reduce((total, budget) => total + budget.maxInputTokens * settings.priceCeilingUsdPerMillionInput / 1_000_000 + budget.maxOutputTokens * settings.priceCeilingUsdPerMillionOutput / 1_000_000, 0);
	const costCeilingPoints = Math.ceil(costUsd * settings.pointsPerUsd);
	return { kind, lineItems: budgets, costCeilingPoints, minimumPoints: minimum, points: Math.max(minimum, costCeilingPoints) };
};

export const settlePoints = (reportedCostUsd: number | null | undefined, kind: string, settings: BillingSettings): number => {
	const minimum = kind === INDEPENDENT_QUOTE ? settings.minimumIndependentEvaluationPoints : settings.minimumEmployerResumePoints;
	return reportedCostUsd == null ? minimum : Math.max(minimum, Math.ceil(reportedCostUsd * settings.pointsPerUsd));
};
